modules/jta/src/main/java/org/jboss/weld/jta/TransactionalObserverNotifier.java

```diff
@@ -78,7 +82,22 @@ public ObserverNotifier create(String contextId, TypeSafeObserverResolver resolv
             for (ObserverMethod<? super T> observer : observers) {
                 deferNotification(event, metadata, observer, notifications);
             }
-            transactionServices.registerSynchronization(new TransactionNotificationSynchronization(notifications));
+            try {
+                transactionServices.registerSynchronization(new TransactionNotificationSynchronization(notifications));
+            } catch (Exception e) {
+                if (e.getCause() instanceof RollbackException || e.getCause() instanceof IllegalStateException) {
+                    List<ObserverMethod<? super T>> filteredObservers = observers.stream()
+                            .filter(observerMethod -> !observerMethod.getTransactionPhase().equals(TransactionPhase.AFTER_SUCCESS)).collect(
+                                    Collectors.toList());
+                    filteredObservers.sort((o1, o2) -> {
```

Comment:
Yes but I guess I need to use Collector again then. This seemed to me more straightforward.